### DISCUSSION TITLE
JWT: How to deal with it if the jwt contains too many roles and is therefore too large?

### DISCUSSION THREAD

**Discussion by npriebe** [keycloak/keycloak#14032] [Q&A] [answered]
### Description

Hello,

In my case, I have a Java Spring backend applications that requests its JWT via Keycloak.
Now I have the problem that my JWT contains a lot of roles, because this token is used to query a lot of applications.
But the token has now exceeded the maximum header size of 8KB due to all the roles.

Is there any way to reduce the size of the JWT while keeping the roles?

**Comment by lexcao** [comment, marked as answer]
Just curious, how many roles have you added?

**Reply by npriebe** [reply]
In this case, my token has roles of, say, 10 different clients. And each client has about 20 roles. 
I need so many roles because this token is used to query the availability and functions of various applications.

Maybe you know a handling for how to deal with something like that?

**Comment by sugilite** [comment]
I feel like this question is more suited to be asked on the Keycloak User Mailing List (https://groups.google.com/d/forum/keycloak-user).

Still some thoughts:

One option would be to only map your roles to the UserInfo endpoint. Your applications then grab the roles from the UserInfo endpoint (you could then even cache the response locally for a certain time while keeping the security implications in mind).

Additionally, you write "because this token is used to query a lot of applications": maybe, depending on your use case, you have too many _realm_ roles which you use for multiple applications. If this is the case you can use client roles and then only map the needed roles for a given client into the access token.

**Comment by schuerg** [comment]
With so many roles, I think it makes sense to fetch the userinfo endpoint after you have a JWT access token, rather than passing the roles as claims inside of the access token.

This way the JWT Access Token stays small, regardless of the number of roles.

See:

- https://www.keycloak.org/docs/latest/securing_apps/index.html#userinfo-endpoint
- https://openid.net/specs/openid-connect-core-1_0.html#UserInfo

**Reply by npriebe** [reply]
Thank you for your answer.
I have come across the possibility several times now. I will try to apply it in my case.

**Reply by pvsreejesh** [reply]
Hi, are you able to solve this issue using the user info end point. I am also facing the same large token size.

**Reply by npriebe** [reply]
Hi, yes I have solved the problem as follows:

In Keycloak under "Client scopes" -> "roles" -> "mappers" -> "client roles" you can determine whether the roles should be added to the "ID token", "Access token" or "Userinfo". I have simply said here that the roles can only be queried under the userinfo endpoint and are no longer added to the access token.
If you then receive your token from the token endpoint, you can use it to query the userinfo endpoint and receive the roles in the userinfo.

Info: If you make the setting via the "Client scope", you can still place the roles for individual clients in the token via a "Client scope" directly on the client.

**Reply by pvsreejesh** [reply]
great ,, thanks !

**Reply by Romain7495** [reply]
Hi, the solution to deal with userinfo endpoint is working well only if you use the role claim in your backend application. If you are sending back the jwt to keycloak authorization server, this is not working if role is not in access token.

cc @schuerg
